tiledb/sm/query/sparse_global_order_reader.cc

```diff
@@ -461,65 +461,54 @@ SparseGlobalOrderReader::compute_result_cell_slab() {
-    std::vector<std::list<ResultTileWithBitmap<uint8_t>>::iterator>&
-        result_tiles_it,
-    std::vector<uint8_t>& result_tile_used,
-    std::priority_queue<ResultCoords, std::vector<ResultCoords>, T>& tile_queue,
-    std::mutex& tile_queue_mutex) {
-  bool found = false;
+template <class CompType>
+tuple<Status, optional<bool>> SparseGlobalOrderReader::add_next_cell_to_queue(
+    GlobalOrderResultCoords& rc,
+    std::vector<TileListIt>& result_tiles_it,
+    TileMinHeap<CompType>& tile_queue) {
+  // Try the next cell in the same tile.
+  auto frag_idx = rc.tile_->frag_idx();
+  bool found = rc.next_cell();
+
+  // If found, update the fragment index, add the cell to the queue and return.
+  if (found) {
+    std::unique_lock<std::mutex> ul(tile_queue_mutex_);
+    read_state_.frag_idx_[frag_idx] = FragIdx(rc.tile_->tile_idx(), rc.pos_);
+    tile_queue.emplace(std::move(rc));
+    return {Status::Ok(), false};
+  }
+
+  // Save the potential tile to delete and increment the tile iterator.
+  auto to_delete = result_tiles_it[frag_idx];
+  result_tiles_it[frag_idx]++;
 
   // Remove the tile from result tiles if it wasn't used at all.
-  if (!result_tile_used[frag_idx]) {
-    auto to_delete = result_tiles_it[frag_idx];
-    to_delete--;
+  if (!rc.tile_->used_) {
     remove_result_tile(frag_idx, to_delete);
   }
 
-  // Try to find a tile.
-  while (!found && result_tiles_it[frag_idx] != result_tiles_[frag_idx].end()) {
-    bool has_bmp = result_tiles_it[frag_idx]->bitmap_.size() > 0;
-    found = !has_bmp;
-    auto tile = &*result_tiles_it[frag_idx];
-
-    // Find a cell that's in the subarray.
-    if (has_bmp) {
-      while (cell_idx < tile->cell_num()) {
-        if (tile->bitmap_.size() == 0 || tile->bitmap_[cell_idx]) {
-          found = true;
-          break;
-        }
+  // Try to find a new tile.
+  if (result_tiles_it[frag_idx] != result_tiles_[frag_idx].end()) {
+    // Find a cell in the current result tile.
+    GlobalOrderResultCoords rc(&*result_tiles_it[frag_idx], 0);
+    found = rc.next_cell();
 
-        cell_idx++;
-      }
-    }
+    // All tiles should at least have one cell available.
+    assert(found == true);
 
-    // There was more tiles in this fragment, insert it in the queue.
-    if (found) {
-      std::unique_lock<std::mutex> ul(tile_queue_mutex);
-      tile_queue.emplace(tile, cell_idx);
-      result_tiles_it[frag_idx]++;
-    } else {
-      // Remove the tile.
-      auto to_delete = result_tiles_it[frag_idx];
-      result_tiles_it[frag_idx]++;
-      remove_result_tile(frag_idx, to_delete);
+    // Insert the cell in the queue.
+    {
+      std::unique_lock<std::mutex> ul(tile_queue_mutex_);
+      tile_queue.emplace(std::move(rc));
     }
-
-    result_tile_used[frag_idx] = false;
-
-    // Once we move to the next tile, the saved cell index doesn't matter.
-    cell_idx = 0;
+    read_state_.frag_idx_[frag_idx] = FragIdx(rc.tile_->tile_idx(), rc.pos_);
   }
 
   if (!found) {
```

Comment:
So if I read well this condition can be true only if the previous `if` 
`if (result_tiles_it[frag_idx] != result_tiles_[frag_idx].end()) {` 
is `false`. So, can we instead add it as an `else` statement for clarity?